### DISCUSSION TITLE
Draw directly to SKSurface?

### DISCUSSION THREAD

**Discussion by bforl** [mono/SkiaSharp#2942] [General]
Assuming I am using the CPU renderer (NOT GPU), i.e the back buffer is in RAM, not GPU memory

Is it possible to get a reference to the buffer Skia is writing to so I can do my own rendering to it?

I'm not looking to get a copy, I want to draw directly to the memory itself? is this possible?

So basically, I want to do something like : 

```
private void skiaView_PaintSurface(object sender, SKPaintSurfaceEventArgs e)
{
	var canvas = e.Surface.Canvas;
	canvas.Clear(SKColors.White);

	// normal rendering
       canvas.DrawLine(....);

       fixed(byte buffer* = canvas.Surface.GetBuffer())
       {
            // write to the actual buffer skia is using
            buffer[0] = ...
       }
}
```

**Comment by ToolmakerSteve** [comment]
I don't know the answer to your question, but it is likely specific to each platform.

I don't know about SkiaSharp, but in older OpenGL frameworks, the solution was typically for you (the app coder) to create the platform-specific object to be used, and pass that to the renderer. But you'd have to be careful how and when you access that object directly.

-----------------
Consider an alternative approach:

* Render to an offscreen bitmap. Then you can do whatever you want.

Also consider having TWO offscreen bitmaps. A "Live" one that you are manipulating, and a "Last" (or "Snapshot") that is ready to be copied to the onscreen Canvas, in its next PaintSurface. An example of this is [C# Data Visualization / Animating Graphics without Blocking the GUI Thread](https://swharden.com/csdv/system.drawing/threading/).

**Reply by bforl** [reply]
Thanks - its funny you should link to Scott Wardens site, because the use case for this is actually ScottPlot!

However as mentioned, this is the CPU renderer, not the GPU renderer. so there is only one 'platform'.

What I was wanting was simply to render to the same back buffer Skia is using at a certain point.

The solution turned out to be very simple (I got help from the Skia Google Group)

```
 public unsafe void CustomRender(SKSurface surface)
 {
     var pixels = surface.PeekPixels();
     uint* buffer= (uint*)pixels.GetPixels();
```

**Reply by ToolmakerSteve** [reply]
Great! I see that opens up interesting possibilities.
